compute_sdk/globus_compute_sdk/sdk/executor.py

```diff
@@ -766,28 +789,44 @@ def shutdown(self, wait=True, *, cancel_futures=False):
                     self,
                     self._task_submitter.name,
                 )
-                try:
-                    # first, drain the queue
-                    while True:
-                        self._tasks_to_send.get(block=False)
-                        self._tasks_to_send.task_done()
-                except queue.Empty:
-                    pass
+
+                if cancel_futures:
+                    try:
+                        while True:
+                            fut, _ = self._tasks_to_send.get(block=False)
+                            fut.cancel()
+                            fut.set_running_or_notify_cancel()
+                            self._tasks_to_send.task_done()
+                    except queue.Empty:
+                        pass
+                else:
+                    while not self._tasks_to_send.empty():
+                        time.sleep(0.1)
                 self._tasks_to_send.put((None, None))  # poison pill for submitter
 
-            if self._result_watcher:
-                self._result_watcher.shutdown(wait=wait, cancel_futures=cancel_futures)
-                self._result_watcher = None
+                if wait:
+                    rw = _RESULT_WATCHERS.get(self.task_group_id)
+                    if rw:
+                        futures = rw.get_open_futures(self)
+                        concurrent.futures.wait(futures.values())
```

Comment:
Right now, if a concurrent call to shutdown is made while one is in the process of waiting for futures, it would return immediately giving the impression that the executor has shutdown. I recommend moving the `self._stopped = True` to the end of this block, so that any other concurrent call to shutdown will not exist early but wait for the shutdown to happen.

Reply:
That's probably the right move, but the `self._stopped` flag also stops new function registrations and task submissions, so a little more work will be needed. Let's do this work in a separate PR.